rfd/0162-machine-id-token-join-method-bot-instance.md

```diff
@@ -0,0 +1,653 @@
+
+// Response for ListBotInstances.
+message ListBotInstancesResponse {
+  // BotInstance that matched the search.
+  repeated BotInstance bot_instances = 1;
+  // Token to retrieve the next page of results, or empty if there are no
+  // more results exist.
+  string next_page_token = 2;
+}
+
+// Request for DeleteBotInstance.
+message DeleteBotInstanceRequest {
+  // The name of the BotInstance to delete.
+  string bot_name = 1;
+  // The unique identifier of the bot instance to delete.
+  string id = 2;
+}
+```
+
+### Changes to the `token` Join Method
+
+As we now have a way to track the generation for a specific Bot instance, we
+can allow multiple Bot instances to be associated with a single Bot. This
+also means that the token no longer needs to be consumed on a join.
+
+However, this does introduce a change in our security guarantees, and without
+additional tooling support and sensible defaults, the change may incentivize end
+users to create long-lived join tokens instead of using a more appropriate join
+method, or automating issuance of short lived tokens.
+
+To this end, we should introduce a per-bot-instance join count limit, and
+configure that to be 1 join by default. This matches today's behavior, and
+will help ensure users do not accidentally create a token that provides more
+access than expected: "infinite use" tokens with massive join limits and/or very
+long TTLs will need to be explicitly specified.
+
+We may additionally want to put hurdles in the way of things like extremely long
+token TTLs. There are legitimate low-security use cases for these, but we could
+introduce a soft limit in `tctl` preventing automatic token creation with TTL
+longer than 7 days, forcing users to manually create the token if they really
+need it to last longer.
+
+Additionally, renewal logic will need to be adjusted to read and update the
+generation counter from the BotInstance rather than the Bot user. We should also
+take care to ensure this counter behaves sensibly even when many bots are
+attempting to join the cluster concurrently. The generation counter today
+already has issues with concurrent joins, and it's even more important to get
+this right when we can expect contention over a single bot instance resource.
+
+### CLI Changes
+
+#### `tbot`
+
+`tbot reset` will be added to allow a `tbot` instance to be reset. This will
+simply clear out any artifacts within the `tbot` storage directory.
+
+In addition, if the bot detects a change in join token or join method, it should
+automatically rotate it's keypair. This will ensure it presents as a fresh
+BotInstance to the AuthServer.
+
+A log message should be output that identifies the linked BotInstance at
+startup and on each heartbeat. This will allow users to easily correlate the
+`tbot` installation with a BotInstance.
+
+#### `tctl`
+
+Commands to list all BotInstances and the BotInstances for a specific Bot should
+be added to the `tctl bots` family:
+
+`tctl bots instances list`
```

Comment:
Do we need `instances`? What about `tctl bots ls`?

Reply:
I think the two are distinct, `tctl bots ls` will continue to work as it does today, showing (effectively) a name + allowed roles. You'd then list all the unique joins of a given bot with `tctl bots instances list --bot <name>`.

There's some overlap, I think we could amend `tctl bots ls` to say, show a unique count of instances per bot, for example.

Reply:
Yeah - that's correct, we'll still want way to list the actual bot identities.